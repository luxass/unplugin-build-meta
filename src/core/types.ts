import type { UnpluginBuildContext, UnpluginContext } from "unplugin";

export type Module = "git";

export interface BuildMetaModule {
  name: string;
  load: (ctx: UnpluginBuildContext & UnpluginContext, id: string) => Promise<string>;
}

export interface BuildMetaOptions {
  /**
   * Which modules to include in the build meta.
   * By default `all` modules are included.
   */
  modules?: (Module | BuildMetaModule)[];
}
