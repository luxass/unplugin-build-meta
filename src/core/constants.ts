export const PLUGIN_NAME = "unplugin-build-meta";
export const NUXT_CONFIG_KEY = "buildMeta";
