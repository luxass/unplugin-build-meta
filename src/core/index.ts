import type { UnpluginFactory, UnpluginInstance } from "unplugin";
import type { BuildMetaModule, BuildMetaOptions } from "./types";
import { createUnplugin } from "unplugin";
import { PLUGIN_NAME } from "./constants";
import gitModule from "./modules/git";
import { runtimeModule } from "./modules/runtime";

export type { BuildMetaOptions };

const PREFIX = "virtual:build-meta/";
const PREFIX_WITH_NULL = `\0${PREFIX}`;

/**
 * A unplugin factory, used by Unplugin to create a new plugin instance.
 */
export const unpluginFactory: UnpluginFactory<BuildMetaOptions | undefined> = (options = {}) => {
  const { git = true, extraModules = [], runtime = true } = options;

  const modules: BuildMetaModule[] = [];

  if (git) {
    modules.push(gitModule);
  }

  if (runtime) {
    modules.push(runtimeModule);
  }

  // Add any extra custom modules
  modules.push(...extraModules);

  return {
    name: PLUGIN_NAME,
    enforce: "pre",
    resolveId(id) {
      if (!id.startsWith(PREFIX)) return;
      id = id.slice(PREFIX.length);
      if (!modules.map((m) => m.name).includes(id)) {
        return;
      }

      return `${PREFIX_WITH_NULL}${id}`;
    },
    loadInclude(id) {
      if (!id.startsWith(PREFIX_WITH_NULL)) return false;
      return modules.map((m) => m.name).includes(id.slice(PREFIX_WITH_NULL.length));
    },
    async load(id) {
      if (!id.startsWith(PREFIX_WITH_NULL)) return;
      id = id.slice(PREFIX_WITH_NULL.length);

      for (const mod of Object.values(modules)) {
        if (id === mod.name) {
          return mod.load(this, id);
        }
      }
    },
  };
};

/**
 * The main unplugin instance.
 */
export const unplugin: UnpluginInstance<BuildMetaOptions | undefined, boolean> = /* #__PURE__ */ createUnplugin(unpluginFactory);

export default unplugin;
