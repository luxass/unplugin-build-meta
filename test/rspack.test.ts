import type { Configuration, Stats } from "@rspack/core";
import { join } from "node:path";
import { rspack as createRspack } from "@rspack/core";
import { assert, describe, expect, it } from "vitest";
import { testdir } from "vitest-testdirs";
import buildMeta from "../src/rspack";

interface RspackResult {
  stats: Stats;
  json: ReturnType<Stats["toJson"]>;
  file: string;
}

async function rspack(config: Configuration, testdirPath: string): Promise<RspackResult> {
  return new Promise((resolve, reject) => {
    const compiler = createRspack({
      mode: "none",
      optimization: {
        minimize: false,
      },
      output: {
        path: join(testdirPath, "dist"),
        filename: "bundle.js",
        libraryTarget: "module",
      },
      target: ["es2020"],
      experiments: {
        outputModule: true,
        rspackFuture: {
          bundlerInfo: {
            force: false,
          },
        },
      },
      ...config,
    });

    compiler.run((err, stats) => {
      if (err) {
        reject(err);
      }

      if (!stats) {
        reject(new Error("rspack stats not available"));
        return;
      }

      const json = stats.toJson();
      const files = json.assetsByChunkName?.main;
      if (!files || !Array.isArray(files) || files[0] == null) {
        reject(new Error("main chunk not found"));
        return;
      }

      const file = files[0];

      resolve({ stats, json, file });
    });
  });
}

describe("handles git metadata", () => {
  it("expect git metadata to be available", async () => {
    const testdirPath = await testdir.from(join(import.meta.dirname, "fixtures/git"));
    expect(testdirPath).toBeDefined();

    const { stats } = await rspack({
      entry: join(testdirPath, "git.js"),
      plugins: [buildMeta({
        modules: ["git"],
      })],
    }, testdirPath);

    assert(stats, "stats should be defined");
    const output = stats.toJson({ source: true }).modules?.[0]?.source;
    expect(output).toBeDefined();
    expect(output).toMatchSnapshot();
  });

  it("expect specific git properties to be importable", async () => {
    const testdirPath = await testdir.from(join(import.meta.dirname, "fixtures/git"));
    expect(testdirPath).toBeDefined();

    const { stats } = await rspack({
      entry: join(testdirPath, "git-specific.js"),
      plugins: [buildMeta({
        modules: ["git"],
      })],
    }, testdirPath);

    const output = stats.toJson({ source: true }).modules?.[0]?.source;
    expect(output).toBeDefined();
    expect(output).toMatchSnapshot();
  });
});
