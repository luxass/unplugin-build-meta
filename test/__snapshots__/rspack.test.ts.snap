// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`handles git metadata > expect git metadata to be available 1`] = `
"import * as git from "virtual:build-meta/git";

console.log(git);
"
`;

exports[`handles git metadata > expect specific git properties to be importable 1`] = `
"import { branch, sha, shortSha } from "virtual:build-meta/git";

console.log({ branch, sha, shortSha });
"
`;
